### ISSUE TITLE
Request to merge Mongoose and Bookshelf Templates and Centralize them

### ISSUE THREAD

**Issue by derrickmehaffy** [strapi/strapi#2912]
- [ ] **I have created my request on the Product Board before I submitted this issue**
- [x] **I have looked at all the other requests on the Product Board before I submitted this issue**

**Please describe your feature request:**

This is a two part feature / framework request.

# Part 1 - Merge the templates

What my thinking is:

We take both sets of templates:
https://github.com/strapi/strapi/tree/master/packages/strapi-generate-api/templates/bookshelf
https://github.com/strapi/strapi/tree/master/packages/strapi-generate-api/templates/mongoose

And merge them together, this way no matter what database you are on (Mongo, MySQL, Postgres, SQLite) you can very easily just change your hook and swap out your DB creds and move between them. There are a number of issues when a user thinks they want to use DB type A, but later on want to switch to DB type b and have to completely regenerate their models to do so (Mongoose to Bookshelf and vice versa).

Each time the model functions are called it will look up what hook type is being used from the `model.settings.json` & the `database.json` and use that "part" of the template. The downside is that yes the template will be much harder to view and understand, but nothing a few comments can't fix :wink: 

# Part 2 - Centralize the templates **with overrides**

To help deduplicate the massive number of files (and make updating easier) we should centralize the config, controllers, and services for all models and if the user wishes to modify them they can just copy the structure.

So basically you would have:
`api/$Model/models/$Model.js` and `api/$Model/models/$Model.settings.json` by default.
Or **if** any of the other files exist, use those instead, so for example:
IF `api/$Model/config/routes.js` exists, use that, else use `$centralLocation/config/routes.js`

The benefit here is if you have 90 models (like I do) I don't need 90 versions of the config/routes.js when I haven't modified them, this only adds to the space (and could increase processing time). Likewise it also makes automated updates that much easier as when a new version is pushed the central locations can be updated automatically.

**Comment by Aurelsicoko** [comment]
Hey, @derrickmehaffy I've created a new card on PB https://portal.productboard.com/strapi/c/55-merge-mongoose-and-bookshelf-templates :)

**Comment by MatheusRV** [comment]
why this is moved to lauched on productboard?

**Comment by derrickmehaffy** [comment]
@MatheusRV because they were merged in the beta with the coreAPI:

https://github.com/strapi/strapi/tree/master/packages/strapi/lib/core-api
